Added new affinity mode for maximum session stickyness.

alexmaret opened the pull request:
**What this PR does / why we need it**:
An alternate implementation to #4482 which keeps backwards compatibility.

**Which issue this PR fixes**: fixes #4475 

**Special notes for your reviewer**:
This implementation keeps backwards compatibility by introducing a new annotation to set the stickyness of a session. The 'balanced' mode (default) uses the old implementation which redistributes some sessions when a deployment is scaled, whereas the 'persistent' mode will offer maximum stickyness and never redistribute a session as long as the target pod is still available.

alexmaret commented:
@ElvinEfendi What do you think of this implementation? It keeps backwards compatibility to the old stickyness mode.

alexmaret commented:
/assign ElvinEfendi

ElvinEfendi commented:
@alexmaret sorry for late responses here. I'll try to review this and the other PR today and provide feedback.

alexmaret commented:
> @alexmaret sorry for late responses here. I'll try to review this and the other PR today and provide feedback.

Thank you, I updated this PR with a proper implementation of the different affinity modes in the form of separate balancer classes. Now the PR is even larger, but I did not really like my previous implementation where I tried to keep the changes at a minimum. This new implementation should be much  easier to understand and easier to maintain.

aledbf commented:
/ok-to-test

ElvinEfendi commented on rootfs/etc/nginx/lua/balancer/sticky.lua:
```diff
@@ -151,30 +130,34 @@ function _M.balance(self)
   ngx_log(INFO, string_format("[%s] nodes have changed for backend %s", self.name, backend.name))
 
+  self.traffic_shaping_policy = backend.trafficShapingPolicy
+  self.alternative_backends = backend.alternativeBackends
```
Moving these to `sync` functions seems to be OK because you call this `sync` function from `new` functions in sticky balancer implementations. However I think this introduces a bug:

Imagine a scenario where list of endpoints changes, alternatives backends get updated but affinity annotations stay unchanged. In this case we will not update `self.alternative_backends` because we will short-circuit in https://github.com/kubernetes/ingress-nginx/pull/4514/files#diff-8f83606fdd5dc8d2406cda71b9609762R154. I think we should move these two lines to above https://github.com/kubernetes/ingress-nginx/pull/4514/files#diff-8f83606fdd5dc8d2406cda71b9609762R148

alexmaret replied:
> Imagine a scenario where list of endpoints changes, alternatives backends get updated but affinity annotations stay unchanged.

@ElvinEfendi 
Actually, is there any good reason to keep the deep_compare at all? I would imagine that this is much more expensive than just updating self.cookie_session_affinity. After all we don't reinitialize anything if that configuration changes, we just assign a variable if it changed.

I will just remove the deep_compare. If you are not happy with that, I can always change it back later.

ElvinEfendi commented on rootfs/etc/nginx/lua/balancer/sticky.lua:
```diff
@@ -86,23 +86,8 @@ local function get_failed_upstreams()
-end
-
 local function should_set_cookie(self)
+
```
this is redundant

ElvinEfendi commented on rootfs/etc/nginx/lua/balancer/sticky_persistent.lua:
```diff
@@ -0,0 +1,56 @@
+
+    if routing_key == nil then
+      local err = string.format("Failed to extract routing key from cookie '%s'!", cookie_value)
+      return nil, err
```
This error is not handled anywhere. Please also make sure test covers this case where we can not extract routing key.

ElvinEfendi commented on rootfs/etc/nginx/lua/balancer/sticky_balanced.lua:
```diff
@@ -0,0 +1,57 @@
+end
+
+function _M.set_routing_key(self, key)
+	self:set_cookie(key)
```
intendentation is off

ElvinEfendi commented on rootfs/etc/nginx/lua/util/nodemap.lua:
```diff
@@ -0,0 +1,120 @@
+-- @tparam {[string]=number} endpoints A table with the node endpoint as a key and its weight as a value.
+-- @tparam[opt] string hash_salt A optional hash salt that will be used to obfuscate the hash key.
+function _M.new(self, endpoints, hash_salt)
+
```
redundant line

ElvinEfendi commented on rootfs/etc/nginx/lua/balancer/sticky_persistent.lua:
```diff
@@ -0,0 +1,56 @@
+end
+
+function _M.set_routing_key(self, key)
+  local value = string.format("%s.%s.%s", ngx.now(), ngx.worker.pid(), key)
```
Why do you need `ngx.now()` and `ngx.worker.pid()`? In `sticky_balanced` we used them to introduce some entropy for better balancing but in `sticky_persistent` I can not think of a reason to keep them.

If we can avoid this we can then just put set/get_routing_key in sticky.lua only and not deal with error handling https://github.com/kubernetes/ingress-nginx/pull/4514/files#r324818550

ElvinEfendi replied:
They key is not designed for that kind of use. I think we should change it to hash-code only.

ElvinEfendi commented on rootfs/etc/nginx/lua/util/nodemap.lua:
```diff
@@ -0,0 +1,120 @@
+local math = require("math")
+local util = require("util")
```
`tablelength` seems to be the only function you use from utils, for better locality (hence perf) you can do

```
local tablelength = require("util").tablelength
```

and use `tablelength` directly

ElvinEfendi commented on rootfs/etc/nginx/lua/util/nodemap.lua:
```diff
@@ -0,0 +1,120 @@
+local math = require("math")
```
same as https://github.com/kubernetes/ingress-nginx/pull/4514/files#r324827563, you only use `random` function from this module

ElvinEfendi commented:
@alexmaret I think this is a good alternative to https://github.com/kubernetes/ingress-nginx/pull/4482. I like the approach here more.

The PR mostly looks good, had only one major feedback what I think is a bug.

alexmaret commented:
@ElvinEfendi Thanks for the review, will look into your findings asap. Just did not have time yet.

aledbf commented:
@alexmaret any chance you can address the comments before next Friday?
We plan to release 0.26 and I want to include this change in the relase.

alexmaret commented:
> @alexmaret any chance you can address the comments before next Friday?
> We plan to release 0.26 and I want to include this change in the relase.

@aledbf 
Yes, I will try to get it done tomorrow or the day after.

alexmaret commented:
/restest

alexmaret commented:
/retest

alexmaret commented:
/retest

alexmaret commented:
@aledbf @ElvinEfendi 
I fixed the review findings.

Not sure why the lualint test case is constantly failing. I fixed the finding in https://github.com/kubernetes/ingress-nginx/commit/c26ab315b8fb838a2fa8337c4d3de0a274647709 but it is still reported as a failure. Even after retesting twice.

alexmaret commented:
/retest

ElvinEfendi commented:
/lgtm

ElvinEfendi commented:
Thanks @alexmaret !
